### PR TITLE
Interpolation

### PR DESCRIPTION
**Pull request by agriggio**
added some basic support for interpolation

### CODE REVIEW

File: pysmt/factory.py [hunk 1]
```diff
@@ -28,25 +28,29 @@
...
 from pysmt.logics import AUTO as AUTO_LOGIC
 from pysmt.logics import most_generic_logic, get_closer_logic
 from pysmt.oracles import get_logic
 
 DEFAULT_SOLVER_PREFERENCE_LIST = ['msat', 'z3', 'cvc4', 'yices', 'picosat', 'bdd']
 DEFAULT_QELIM_PREFERENCE_LIST = ['z3', 'msat_fm', 'msat_lw']
+DEFAULT_INTERPOLATION_PREFERENCE_LIST = ['msat', 'z3']
 DEFAULT_LOGIC = QF_UFLIRA
 DEFAULT_QE_LOGIC = LRA
+DEFAULT_INTERPOLATION_LOGIC = QF_UFLRA
```
Comment (mikand) [inline pysmt/factory.py]: Why QF_UFLRA?  Is there a reason why QF_UFLIRA is not the default? (is it even supported?)

File: pysmt/solvers/interpolation.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    def __enter__(self):
+        """ Manage entering a Context (i.e., with statement) """
+        return self
+
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """ Manage exiting from Context (i.e., with statement)
+
+        The default behaviour is to explicitely destroy the qelim to free
```
Comment (mikand) [inline pysmt/solvers/interpolation.py]: Copy/paste leftover

File: pysmt/solvers/msat.py [hunk 3]
```diff
@@ -904,3 +905,80 @@ class MSatLWQuantifierEliminator(MSatQuantifierEliminator):
...
+        mathsat.msat_destroy_env(self.msat_env)
+
+
+    def _check_logic(self, formulas):
+        for f in formulas:
+            logic = get_logic(f, self.environment)
+            ok = any(logic <= l for l in self.LOGICS)
+            if not ok:
+                raise NotImplementedError(
+                    "Logic not supported by Z3 inteprolation."
```
Comment (mikand) [inline pysmt/solvers/msat.py]: Typo: this is MathSAT! :)

File: pysmt/solvers/z3.py [hunk 4]
```diff
@@ -491,3 +492,52 @@ def eliminate_quantifiers(self, formula):
                                           expression=res)
 
         return pysmt_res
+
+
+class Z3Interpolator(Interpolator):
+
+    LOGICS = [QF_AUFLIA, QF_UFLRA]
```
Comment (mikand) [inline pysmt/solvers/z3.py]: Currently, PySMT does not support arrays, so this should be QF_UFLIA. (At least, we did not included non-representable logics in other solvers / quantelims)

File: pysmt/test/test_interpolation.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+from pysmt.test import (TestCase, skipIfSolverNotAvailable)
+from pysmt.exceptions import (SolverReturnedUnknownResultError, \
+                              NoSolverAvailableError)
+from pysmt.logics import LRA, LIA, UFLIRA
+
+
+class TestInterpolation(TestCase):
+    def test_selection(self):
+        with self.assertRaises(NoSolverAvailableError):
+            QuantifierEliminator(logic=UFLIRA)
```
Comment (mikand) [inline pysmt/test/test_interpolation.py]: Should be Interpolator

File: pysmt/test/test_interpolation.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+        with self.assertRaises(NoSolverAvailableError):
+            QuantifierEliminator(name="nonexistent")
```
Comment (mikand) [inline pysmt/test/test_interpolation.py]: Should be Interpolator

File: pysmt/test/test_interpolation.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    def _bool_example(self, itp, binary):
+        # Bool Example
+        x, y, z = Symbol("bx"), Symbol("by"), Symbol("bz")
+
+        a = And(x, Not(y))
+        b = And(Implies(z, y), z)
+
+        if binary:
+            i = itp.binary_interpolant(a, b)
+            self.assertTrue(i is not None)
```
Comment (mikand) [inline pysmt/test/test_interpolation.py]: This line is useless (This is checked at line 84)

File: pysmt/factory.py (hunk 1 above)
Comment (agriggio) [inline pysmt/factory.py]: MathSAT at least can produce wrong interpolants for QF_LIRA, and AFAIK also Z3 is incomplete with integers

Comment (mikand) [inline pysmt/factory.py]: OK

File: pysmt/solvers/interpolation.py (hunk 2 above)
Comment (agriggio) [inline pysmt/solvers/interpolation.py]: ? you mean the docstring is wrong? I'll fix that

**Comment by mikand** [comment]
COOL! Thanks @agriggio !

the only point is that this PR includes an unrelated commit (infix variables). I will cherry-pick your commit in a fresh branch from master, fix the typos (unless you want to do it) and accept it in a separate pull-request.

**Comment by agriggio** [comment]
I'm fixing the typos
I noticed the spurious commit, but wasn't able to get rid of it... sorry, my git knowledge is limited :-/

**Comment by agriggio** [comment]
fixed typos in 0bb214b8d3a33ff940d3669581865f00fbbcd10d

**Comment by mikand** [comment]
Closing this PR. I created a clean one (#104) cherry-picking @agriggio's commits

**Pull request closed without merging**